prometheus/echoprometheus/prometheus.go

```diff
@@ -0,0 +1,458 @@
+
+	// AfterNext is callback that is executed after next middleware/handler returns. Useful for case when you have own
+	// metrics that need incremented/observed.
+	AfterNext func(c echo.Context, err error)
+
+	timeNow func() time.Time
+}
+
+type LabelValueFunc func(c echo.Context, err error) string
+
+// HandlerConfig contains the configuration for creating HTTP handler for metrics.
+type HandlerConfig struct {
+	// Gatherer sets the prometheus.Gatherer instance the middleware will use when generating the metric endpoint handler.
+	// Defaults to: prometheus.DefaultGatherer
+	Gatherer prometheus.Gatherer
+}
+
+// PushGatewayConfig contains the configuration for pushing to a Prometheus push gateway.
+type PushGatewayConfig struct {
+	// PushGatewayURL is push gateway URL in format http://domain:port
+	PushGatewayURL string
+
+	// PushInterval in ticker interval for pushing gathered metrics to the Gateway
+	// Defaults to: 1 minute
+	PushInterval time.Duration
+
+	// Gatherer sets the prometheus.Gatherer instance the middleware will use when generating the metric endpoint handler.
+	// Defaults to: prometheus.DefaultGatherer
+	Gatherer prometheus.Gatherer
+
+	// ErrorHandler is function that is called when errors occur. When callback returns error StartPushGateway also returns.
+	ErrorHandler func(err error) error
+
+	// ClientTransport specifies the mechanism by which individual HTTP POST requests are made.
+	// Defaults to: http.DefaultTransport
+	ClientTransport http.RoundTripper
+}
+
+// NewHandler creates new instance of Handler using Prometheus default registry.
+func NewHandler() echo.HandlerFunc {
+	return NewHandlerWithConfig(HandlerConfig{})
+}
+
+// NewHandlerWithConfig creates new instance of Handler using given configuration.
+func NewHandlerWithConfig(config HandlerConfig) echo.HandlerFunc {
+	if config.Gatherer == nil {
+		config.Gatherer = prometheus.DefaultGatherer
+	}
+	h := promhttp.HandlerFor(config.Gatherer, promhttp.HandlerOpts{})
+
+	if r, ok := config.Gatherer.(prometheus.Registerer); ok {
+		h = promhttp.InstrumentMetricHandler(r, h)
+	}
+
+	return func(c echo.Context) error {
+		h.ServeHTTP(c.Response(), c.Request())
+		return nil
+	}
+}
+
+// NewMiddleware creates new instance of middleware using Prometheus default registry.
+func NewMiddleware(subsystem string) echo.MiddlewareFunc {
+	return NewMiddlewareWithConfig(MiddlewareConfig{Subsystem: subsystem})
+}
+
+// NewMiddlewareWithConfig creates new instance of middleware using given configuration.
+func NewMiddlewareWithConfig(config MiddlewareConfig) echo.MiddlewareFunc {
+	mw, err := config.ToMiddleware()
+	if err != nil {
+		panic(err)
+	}
+	return mw
+}
+
+// ToMiddleware converts configuration to middleware or returns an error.
+func (conf MiddlewareConfig) ToMiddleware() (echo.MiddlewareFunc, error) {
+	if conf.timeNow == nil {
+		conf.timeNow = time.Now
+	}
+	if conf.Subsystem == "" {
+		conf.Subsystem = defaultSubsystem
+	}
+	if conf.Registerer == nil {
```

Comment:
I see this pattern a lot and I'd like to propose an alternative - it involves adding this external library but the library has a pretty small dependency footprint https://github.com/imdario/mergo

Define a struct of defaults and merge it (the implementation uses reflect)
```
defaultConfig := MiddlewareConfig{
   Subsystem: defaultSubsystem,
   timeNow: time.Now,
   HistogramOptsFunc: func(blah blah blah),
}
err := mergo.Merge(&defaults, conf, mergo.WithOverride)
if err != nil { panic(err) }
conf = defaults
```

Things I find appealing about this are
 - it makes the code much briefer
 - you don't have to reason about what the zero-value is for anything because the type system (reflect package) handles that question
 - I find it easier to read


But I understand the resistance to adding dependencies as well as using reflect (which mergo uses heavily in its implementation)

Reply:
These public "defaults" objects are not that useful as looking middleware code separately it seems. 

I do not think I have ever felt the need to have these kind objects - most of the time you can get away just looking if field is nil or default value and replace it with something appropriate in middleware code. Only thing you would need is to fill field that interest you and need non-default values.  It looks eloquent but in real world it just adds complexity. Go is this stupidly simple language and things like that are somewhat against the Ethos of Go.

NB: not all code I produce follow the Ethos but I am simple man that often fails to see his own faults.